tokio/src/io/async_read.rs

```diff
@@ -57,37 +58,17 @@ pub trait AsyncRead {
     ) -> Poll<io::Result<()>>;
 }
 
-macro_rules! deref_async_read {
-    () => {
-        fn poll_read(
-            mut self: Pin<&mut Self>,
-            cx: &mut Context<'_>,
-            buf: &mut ReadBuf<'_>,
-        ) -> Poll<io::Result<()>> {
-            Pin::new(&mut **self).poll_read(cx, buf)
-        }
-    };
-}
-
-impl<T: ?Sized + AsyncRead + Unpin> AsyncRead for Box<T> {
-    deref_async_read!();
-}
-
-impl<T: ?Sized + AsyncRead + Unpin> AsyncRead for &mut T {
-    deref_async_read!();
-}
-
-impl<P> AsyncRead for Pin<P>
+impl<T> AsyncRead for T
 where
-    P: DerefMut + Unpin,
-    P::Target: AsyncRead,
+    T: DerefPinMut,
+    <T as Deref>::Target: AsyncRead,
 {
```

Comment:
This kind of change makes me very nervous breaking-change wise, since it is introducing a blanket impl for a type that did not previously have one, and blanket impls are prone to errors regarding overlapping impls.